biomdi/fingerimage/src/mkfir/mkfir.1
.\""
.Dd August 15, 2006
.Dt MKFIR 1  
.Os Mac OS X       
.Sh NAME
.Nm mkfir
.Nd Create Finger Image Format Records
.Sh SYNOPSIS
.Nm
.Fl h
.Ar hdrinfo
.Fl f
.Ar fivrinfo
.Oo Fl to Ar type Oc
.Fl o
.Ar outfile
.Op Fl p
.Pp
.Sh DESCRIPTION
The
.Nm
command is used to create ANSI or ISO compliant Finger Image-Based
records from a set of ASCII text files and images. The resulting
records are written to the output file.
.Pp
The options are as follows:
.Bl -tag -width -indent
.It Fl h\ \&hdrinfo
Specifies the file containing general record header info, single line,
in this form:
.Bl -tag -width "Image Compression Algorithm" -compact
.It Cm Format ID
Char ("FIR")
.It Cm Version Number
Char ("010")
.It Cm Record Length
Decimal, not currently used
.It Cm CBEFF Product ID
Hex
.It Cm Capture Device ID
Hex
.It Cm Image Acquisition Level
Decimal
.It Cm Number of Fingers/Palms
Decimal
.It Cm Scale Units
Decimal
.It Cm Scan Resolution (horiz)
Decimal
.It Cm Scan Resolution (vert)
Decimal
.It Cm Image Resolution (horiz)
Decimal
.It Cm Image Resolution (vert)
Decimal
.It Cm Pixel Depth
Decimal
.It Cm Image Compression Algorithm
Decimal
.It Cm Reserved
Decimal
.El
.Pp
.It Fl f\ \&fivrinfo
Specifies the file containing information for the Finger Image View Record(s)
(FIVR). The FIVR
contains all the information for a single finger image, including the name
of the raster image file.
The FIVR input file contains pairs lines. The first line of each pair
contains the data for the finger image header record in this form:
.Bl -tag -width "Horizontal Line Length" -compact
.It Cm Length
Decimal, not currently used
.It Cm Finger/Palm Position
Decimal
.It Cm Count of Views
Decimal
.It Cm View Number
Decimal
.It Cm Image Quality
Decimal
.It Cm Impression Type
Decimal
.It Cm Horizontal Line Length
Decimal
.It Cm Vertical Line Length
Decimal
.It Cm Reserved
Decimal
.El
.Pp
The second line of the pair contains the name of the raster image file.
.Pp
.It Fl to\ \&type
Specifies the output file type.
.Pp
Valid types for the output file are:
.Bl -tag -width "ANSI" -compact
.It Cm ANSI
ANSI/INCITS 381-2004 format;
.It Cm ISO
ISO/IEC 19794-4:2005 format.
.El
The default output type is ANSI.
.It Fl o
specifies the name of the output file.
.Pp
.It Fl p
causes the created record(s) to be printed to the screen.
.El
.Pp
The record length field of the header and the length field of the FIVR
are read by
.Nm
but not used as these field values are calculated.
.Sh EXAMPLES
mkfir -h hdr.txt -f fivr.txt -o fir.raw
.Pp
Produces a file containing the finger image record(s) in ANSI format.
.Pp
mkfir -h hdr.txt -f fivr.txt -to ISO -o fir.raw
.Pp
Produces a file containing the finger image record(s) in ISO format.
.Sh FILES
Example header file:
.Bd -literal -compact
FIR 010 0 0x0001000f 0x0000 31 2 1 500 500 500 500 8 3 0
.Ed
.Pp
Example FIVR file:
.Bd -literal -compact
0 7 1 1 3 0 368 380 0
lfing.jpg
0 2 1 1 3 0 368 380 0
rfing.jpg
.Ed
.Sh SEE ALSO
.Xr prfir 1 .
.Sh STANDARDS
``Finger Image-Based Data Interchange Format'', ANSI/INCITS 381-2004,
May 13, 2004.
.Pp
``Information technology - Biometrita interchange formats - Part 4:
Finger image data'', ISO/IEC 19794-4:2004, 2005-06-01.
.Sh HISTORY
Created August 15th, 2006 by NIST.
.Pp
Added support for ISO format, August 28 2008 by NIST.


biomdi/fingerimage/src/mkfir/mkfir.c
/*
* This software was developed at the National Institute of Standards and
* Technology (NIST) by employees of the Federal Government in the course
* of their official duties. Pursuant to title 17 Section 105 of the
* United States Code, this software is not subject to copyright protection
* and is in the public domain. NIST assumes no responsibility  whatsoever for
* its use by other parties, and makes no guarantees, expressed or implied,
* about its quality, reliability, or any other characteristic.
*/
/******************************************************************************/
/* This program uses the ANSI/INCITS Finger Image Record library to           */
/* create a file containing a Finger Image-Base record from on data           */
/* contained in several text files.                                           */
/*                                                                            */
/* Parameters to this program:                                                */
/*   -h <header>      : The name of the header file                           */
/*   -f <fih>         : The name of the finger image header file including    */
/*                      the name of a raster file; multiple instances are     */
/*                      allowed                                               */
/*   -p               : Print the entire FIR to stdout                        */
/*                                                                            */
/******************************************************************************/

/* Needed by the GNU C libraries for Posix and other extensions */
#define _XOPEN_SOURCE	1

#include <sys/param.h>
#include <sys/queue.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <biomdi.h>
#include <biomdimacro.h>
#include <fir.h>

int total_length = 0;

static void
usage()
{
	fprintf(stderr,
	    "usage: mkfir -h <headerfile> -f <fivrfile> [-p] [-to <type>]"
	    " -o <datafile>\n"
	    "\twhere -h, -f, and -o are required\n"
	    "\t -to <type> is one of ISO | ANSI\n");
	exit (EXIT_FAILURE);
}

/******************************************************************************/
/* Load the FIR header info from a text file and set the fields in the finger */
/* image record (FIR).  This is the general record header.                    */
/* The record length will be calculated; value in the file is ignored         */
/******************************************************************************/
int
load_hdr(FILE *fp, struct finger_image_record *fir)
{
	unsigned short eqpt;
	unsigned int cbeff_id;

	// This scan follows the layout in Table 2 of 381-2004
	if (fscanf(fp, "%3c %3c %llu %x %hx %hu %hhu %hhu %hu %hu %hu %hu %hhu %hhu %hu",
	    fir->format_id,
	    fir->spec_version,
	    &fir->record_length,
	    &cbeff_id,
	    &eqpt,
	    &fir->image_acquisition_level,
	    &fir->num_fingers_or_palm_images,
	    &fir->scale_units,
	    &fir->x_scan_resolution,
	    &fir->y_scan_resolution,
	    &fir->x_image_resolution,
	    &fir->y_image_resolution,
	    &fir->pixel_depth,
	    &fir->image_compression_algorithm,
	    &fir->reserved) != 15)
		return (READ_ERROR);
	
	fir->product_identifier_owner = (cbeff_id & HDR_PROD_ID_OWNER_MASK) >>
	    HDR_PROD_ID_OWNER_SHIFT;
	fir->product_identifier_type = cbeff_id & HDR_PROD_ID_TYPE_MASK;
	fir->scanner_id = eqpt & HDR_SCANNER_ID_MASK;
	fir->compliance = (eqpt & HDR_COMPLIANCE_MASK) >> HDR_COMPLIANCE_SHIFT;

	// NULL-terminate the format ID
	fir->format_id[3] = 0x00;

	// NULL-terminate the version number
	fir->spec_version[3] = 0x00;

	if (fir->format_std == FIR_STD_ANSI)
		total_length += FIR_ANSI_HEADER_LENGTH;
	else
		total_length += FIR_ISO_HEADER_LENGTH;
	return (READ_OK);
}

/******************************************************************************/
/* Load a single finger image view record from a text file and add to the     */
/* finger image record.                                                       */
/*                                                                            */
/******************************************************************************/
int
load_fivr(FILE *fp, struct finger_image_record *fir)
{
	struct finger_image_view_record *fivr;
	int ret;
	char filename[MAXPATHLEN];
	char *buf;
	struct stat sb;
	FILE *image_fp;

	if (new_fivr(&fivr) < 0)
		ALLOC_ERR_RETURN("Finger Image View Record");

	// This scan follows the layout of Table 4 in 381-2004
	ret = fscanf(fp, "%u %hhu %hhu %hhu %hhu %hhu %hu %hu %hhu",
		&fivr->length,
		&fivr->finger_palm_position,
		&fivr->count_of_views,
		&fivr->view_number, 
		&fivr->quality,
		&fivr->impression_type,
		&fivr->horizontal_line_length,
		&fivr->vertical_line_length,
		&fivr->reserved);
	if (ret == EOF)
		return (READ_EOF);
	else if (ret != 9)
		return (READ_ERROR);

	if (fscanf(fp, "%s", filename) != 1)
		return (READ_ERROR);

	// Even though we read the length from the header, we set it to
	// the correct size here.
	fivr->length = FIVR_HEADER_LENGTH;

	if (stat(filename, &sb) == 0) {
		if ((image_fp = fopen(filename, "rb")) == NULL) {
			ERRP ("Could not read image file %s", filename);
		}
		else {
			buf = (char *)malloc(sb.st_size);
			if (buf == NULL) {
				ERRP("Could not allocate memory to read image");
			}
			else {
				if (fread(buf, 1, sb.st_size, image_fp) <= 0) {
					ERRP("Could not read image file %s",
					    filename);
				} else {
					fivr->length += sb.st_size;
					add_image_to_fivr(buf, fivr);
				}
			}
		}
	} else {
		ERRP("Could not locate image file %s", filename);
	}

	add_fivr_to_fir(fivr, fir);

	total_length += fivr->length;

	return (READ_OK);
}

int
main(int argc, char *argv[])
{
	FILE *out_fp = NULL;	// for the output file
	FILE *hdr_fp = NULL;	// the header input file
	FILE *fivr_fp = NULL;	// the finger image view record files
	int h_opt, f_opt, o_opt, to_opt, p_opt;
	struct finger_image_record *fir;
	struct stat sb;
	char ch;
	int ret, exit_code;
	int out_type;
	char pm;

	exit_code = EXIT_SUCCESS;
	h_opt = f_opt = o_opt = to_opt = p_opt = 0;
	out_type = FIR_STD_ANSI;
	while ((ch = getopt(argc, argv, "h:f:o:t:p")) != -1) {
		switch (ch) {
			case 'h':
				if ((hdr_fp = fopen(optarg, "r")) == NULL)
					OPEN_ERR_EXIT(optarg);
				h_opt = 1;
				break;
			case 'f':
				if ((fivr_fp = fopen(optarg, "r")) == NULL)
					OPEN_ERR_EXIT(optarg);
				f_opt = 1;
				break;
			case 't':
				pm = *(char *)optarg;
				switch (pm) {
					case 'o':
						out_type = fir_stdstr_to_type(
						    argv[optind]);
						if (out_type < 0)
							usage();
						optind++;
						to_opt++;
						break;
					default:
						usage();
						break;	/* not reached */
				}
				break;
			case 'o':
				if (stat(optarg, &sb) == 0)
					ERR_EXIT("File '%s' exists, remove it first.\n", optarg);

				if ((out_fp = fopen(optarg, "wb")) == NULL)
					OPEN_ERR_EXIT(optarg);
				o_opt = 1;
				break;
			case 'p':
				p_opt = 1;
				break;
			default:
				usage();
				break;
		}
	}
	if ((h_opt && f_opt && o_opt) == 0)
		usage();

	// Read in the file containing the header information
	if (new_fir(out_type, &fir) < 0)
		ALLOC_ERR_EXIT("Finger Image Record (general header)");

	if (load_hdr(hdr_fp, fir) != READ_OK) {
		exit_code = EXIT_FAILURE;
		ERR_OUT("Could not read header");
	}

	// Read in each Finger Image View Record add to the Finger Image Record
	do {
		ret = load_fivr(fivr_fp, fir);
		if (ret == READ_ERROR) {
			exit_code = EXIT_FAILURE;
			ERRP("Reading image view record");
			goto err_out;
		}
	} while (ret == READ_OK);

	fir->record_length = total_length;

	// Validate the Finger Image Record
	if (validate_fir(fir) != VALIDATE_OK) {
		fprintf(stdout, "Finger Image Record is invalid.\n");
	} else {
		fprintf(stdout, "Finger Image Record is valid.\n");
	}

	if (p_opt)
		print_fir(stdout, fir);

	// Write out the FIR
	if (write_fir(out_fp, fir) != WRITE_OK) {
		fprintf(stderr, "Error writing the Finger Image Record\n");
	} else {
		fprintf(stdout, "Finger Image Record written.\n");
	}

err_out:
	free_fir(fir);

	if (o_opt) 
		fclose(out_fp);
	if (h_opt) 
		fclose(hdr_fp);

	if (f_opt)
		fclose(fivr_fp);

	exit(exit_code);
}
